app/controllers/materials_controller.rb

```diff
@@ -0,0 +1,221 @@
+require 'rubygems'
```

Comment:
I think most of the private methods could be moved into some sort of FileHandler class.